services/src/main/java/org/keycloak/cache/DefaultAlternativeLookupProvider.java

```diff
@@ -0,0 +1,78 @@
+package org.keycloak.cache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.IdentityProviderModel;
+import org.keycloak.models.KeycloakSession;
+
+import java.util.List;
+import java.util.Map;
+
+public class DefaultAlternativeLookupProvider implements AlternativeLookupProvider {
+
+    private Cache<String, String> lookupCache;
+
+    public DefaultAlternativeLookupProvider(Cache<String, String> lookupCache) {
+        this.lookupCache = lookupCache;
+    }
+
+    public String lookupIdentityProviderAliasFromIssuer(KeycloakSession session, String issuerUrl) {
+        String alternativeKey = ComputedKey.computeKey(session.getContext().getRealm().getId(), "idp", issuerUrl);
+
+        String idpAlias = lookupCache.getIfPresent(alternativeKey);
+        if (idpAlias != null) {
+            IdentityProviderModel idp = session.identityProviders().getByAlias(idpAlias);
+            if (idp != null && issuerUrl.equals(idp.getConfig().get(IdentityProviderModel.ISSUER))) {
+                return idpAlias;
+            } else {
+                lookupCache.invalidate(alternativeKey);
+            }
+        }
+
+        idpAlias = session.identityProviders().getAllStream()
+                .filter(i -> issuerUrl.equals(i.getConfig().get(IdentityProviderModel.ISSUER)))
+                .map(IdentityProviderModel::getAlias).findFirst().orElse(null);
+        if (idpAlias != null) {
+            lookupCache.put(alternativeKey, idpAlias);
+        }
+        return idpAlias;
+    }
+
+    public String lookupClientIdFromClientAttributes(KeycloakSession session, Map<String, String> attributes) {
```

Comment:
Can we not make this `public ClientModel lookupClientFromClientAttributes(KeycloakSession session, Map<String, String> attributes) {`?

That way we can avoid two calls to `session.clients().getClientByClientId` on a hit.